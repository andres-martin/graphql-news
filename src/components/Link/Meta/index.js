/* global localStorage */
/* eslint-disable no-script-url */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import moment from 'moment';
import { library } from '@fortawesome/fontawesome-svg-core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faBookmark, faClock, faComment, faEye, faTrash } from '@fortawesome/free-solid-svg-icons';
import gql from 'graphql-tag';
import { Query } from 'react-apollo';

import DestroyModal from '../../Modal/Destroy';
import { AUTH_TOKEN } from '../../../constants';

library.add(faBookmark, faClock, faComment, faEye, faTrash);

const GET_CLIENT_USER = gql`
  query GetUser {
    User @client {
      _id
      username
    }
  }
`;

export default class Meta extends Component {
  constructor(props) {
    super(props);

    this.state = {
      showModal: false,
      user: null,
      token: localStorage.getItem(AUTH_TOKEN),
    };

    this.showModal = this.showModal.bind(this);
    this.dismissModal = this.dismissModal.bind(this);
  }

  static propTypes = {
    author: PropTypes.string.isRequired,
    authorId: PropTypes.string.isRequired,
    commentsLength: PropTypes.number.isRequired,
    createdAt: PropTypes.string.isRequired,
    _id: PropTypes.string.isRequired,
  };

  showModal() {
    this.setState({
      showModal: true,
    });
  }

  dismissModal() {
    this.setState({
      showModal: false,
    });
  }

  render() {
    const { _id, author, commentsLength, createdAt, authorId } = this.props;
    const { showModal } = this.state;

    return (
      <div className="meta-wrapper">
        <ul className="meta-list">
          <li className="meta-item">
            <span style={{ marginRight: '2px' }}>by</span>
            <span className="username">{author}</span>
          </li>
          <li className="meta-item">
            <FontAwesomeIcon className="fa-icon" icon="clock" />
            <span>{moment(createdAt, '{YYYY} MM-DDTHH:mm:ss SSS [Z] A').fromNow()}</span>
          </li>
          <li className="meta-item">
            <Link to={`/link/${_id}`}>
              <FontAwesomeIcon className="fa-icon" icon="comment" />
              <span>
                {commentsLength} {commentsLength !== 1 ? 'comments' : 'comment'}
              </span>
            </Link>
          </li>
          <Query query={GET_CLIENT_USER}>
            {({ data }) => {
              const { token } = this.state;
              const { User } = data;
              const isAuthenticated = token && User._id && User.username;

              if (isAuthenticated && authorId === User._id) {
                return (
                  <li className="meta-item">
                    <button className="admin-delete-button" onClick={() => this.showModal()}>
                      <FontAwesomeIcon className="fa-icon" icon="trash" />
                      <span>Delete</span>
                    </button>

                    {showModal && isAuthenticated ? (
                      <DestroyModal
                        id={_id}
                        dismissModal={this.dismissModal}
                        isActive={this.state.showModal}
                      />
                    ) : null}
                  </li>
                );
              }

              return <React.Fragment />;
            }}
          </Query>
        </ul>
      </div>
    );
  }
}
