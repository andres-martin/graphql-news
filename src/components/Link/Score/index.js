import React from 'react';
import PropTypes from 'prop-types';

import UpvoteButton from './upvoteButton';
import DownvoteButton from './downvoteButton';
import ScoreCount from './scoreCount';

const Score = ({ _id, score }) => (
  <div className="score-wrapper">
    <UpvoteButton _id={_id} />
    <ScoreCount _id={_id} score={score} />
    <DownvoteButton _id={_id} />
  </div>
);

Score.propTypes = {
  _id: PropTypes.string.isRequired,
  score: PropTypes.number.isRequired,
};

export default Score;
