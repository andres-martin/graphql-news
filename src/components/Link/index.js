import React from 'react';
import PropTypes from 'prop-types';

import ScoreHOC from './Score';
import Meta from './Meta';

const Link = ({ _id, author, authorId, url, description, commentsLength, createdAt, score }) => (
  <div className="link-wrapper">
    <ScoreHOC _id={_id} score={score} />
    <div
      className="link-content-wrapper"
      style={{ display: 'inline-flex', flexDirection: 'column' }}
    >
      <div className="link-title-wrapper">
        <a className="link-title" href={url}>
          {description.length > 60 ? `${description.substring(0, 60)}...` : description}
        </a>
        <a className="link-portal" href={url}>
          ({url.replace(/^(?:https?:\/\/)?(?:www\.)?/i, '').split('/')[0]})
        </a>
      </div>
      <Meta
        _id={_id}
        author={author}
        authorId={authorId}
        commentsLength={commentsLength}
        createdAt={createdAt}
      />
    </div>
  </div>
);

Link.propTypes = {
  _id: PropTypes.string.isRequired,
  author: PropTypes.string.isRequired,
  authorId: PropTypes.string.isRequired,
  createdAt: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  commentsLength: PropTypes.number.isRequired,
  score: PropTypes.number.isRequired,
};

export default Link;
