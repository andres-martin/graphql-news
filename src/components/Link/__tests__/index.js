import React from 'react';
import TestRenderer from 'react-test-renderer';
import { MockedProvider } from 'react-apollo/test-utils';
import { MemoryRouter } from 'react-router';

import Link from '../index';

describe('the <Link /> component', async () => {
  test('should look like this', async () => {
    const tree = TestRenderer.create(
      <MockedProvider>
        <MemoryRouter>
          <Link
            _id="5c7e92d9b7e6d953fb9f35e9"
            author="example"
            authorId="5c786210614c53755f6de7d2"
            url="https://google.com"
            description="Google"
            commentsLength={25}
            score={10}
            createdAt="{2019} 03-05T10:19:35 665 Z AM"
          />
        </MemoryRouter>
      </MockedProvider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
