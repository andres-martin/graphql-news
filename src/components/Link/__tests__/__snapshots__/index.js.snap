// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the <Link /> component should look like this 1`] = `
<div
  className="link-wrapper"
>
  <div
    className="score-wrapper"
  >
    <button
      className="upvote-button"
      onClick={[Function]}
    >
      <svg
        aria-label="up-arrow"
        className="upvote-up-arrow"
        height="7"
        id="svg-up-arrow"
        role="img"
        version="1.1"
        viewBox="0 0 11 7"
        width="11"
      >
        <path
          d="m.202 5.715c-.367.417-.217.755.329.755h9.438c.549 0 .702-.33.338-.742l-4.41-4.985c-.363-.41-.947-.412-1.322.013l-4.373 4.96"
        />
      </svg>
    </button>
    <span
      className="score-count"
    >
      10
    </span>
    <button
      className="downvote-button"
      onClick={[Function]}
    >
      <svg
        aria-label="down-arrow"
        className="downvote-down-arrow"
        height="7"
        id="svg-up-arrow"
        role="img"
        version="1.1"
        viewBox="0 0 11 7"
        width="11"
      >
        <path
          d="m.202 5.715c-.367.417-.217.755.329.755h9.438c.549 0 .702-.33.338-.742l-4.41-4.985c-.363-.41-.947-.412-1.322.013l-4.373 4.96"
        />
      </svg>
    </button>
  </div>
  <div
    className="link-content-wrapper"
    style={
      Object {
        "display": "inline-flex",
        "flexDirection": "column",
      }
    }
  >
    <div
      className="link-title-wrapper"
    >
      <a
        className="link-title"
        href="https://google.com"
      >
        Google
      </a>
      <a
        className="link-portal"
        href="https://google.com"
      >
        (
        google.com
        )
      </a>
    </div>
    <div
      className="meta-wrapper"
    >
      <ul
        className="meta-list"
      >
        <li
          className="meta-item"
        >
          <span
            style={
              Object {
                "marginRight": "2px",
              }
            }
          >
            by
          </span>
          <span
            className="username"
          >
            example
          </span>
        </li>
        <li
          className="meta-item"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-clock fa-w-16 fa-icon"
            data-icon="clock"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm57.1 350.1L224.9 294c-3.1-2.3-4.9-5.9-4.9-9.7V116c0-6.6 5.4-12 12-12h48c6.6 0 12 5.4 12 12v137.7l63.5 46.2c5.4 3.9 6.5 11.4 2.6 16.8l-28.2 38.8c-3.9 5.3-11.4 6.5-16.8 2.6z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
          <span>
            6 days ago
          </span>
        </li>
        <li
          className="meta-item"
        >
          <a
            href="/link/5c7e92d9b7e6d953fb9f35e9"
            onClick={[Function]}
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-comment fa-w-16 fa-icon"
              data-icon="comment"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={Object {}}
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 32C114.6 32 0 125.1 0 240c0 49.6 21.4 95 57 130.7C44.5 421.1 2.7 466 2.2 466.5c-2.2 2.3-2.8 5.7-1.5 8.7S4.8 480 8 480c66.3 0 116-31.8 140.6-51.4 32.7 12.3 69 19.4 107.4 19.4 141.4 0 256-93.1 256-208S397.4 32 256 32z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
            <span>
              25
               
              comments
            </span>
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
